// @vscode/test-web/out/browser/amd/main.js
define("vscode-web-browser-main", ["require", "exports", "./workbench.api"], function (require, exports, workbench_api_1) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    class WorkspaceProvider {
        workspace;
        payload;
        static QUERY_PARAM_EMPTY_WINDOW = 'ew';
        static QUERY_PARAM_FOLDER = 'folder';
        static QUERY_PARAM_WORKSPACE = 'workspace';
        static QUERY_PARAM_PAYLOAD = 'payload';
        static create(config) {
            let foundWorkspace = false;
            let workspace;
            let payload = Object.create(null);
            const query = new URL(document.location.href).searchParams;
            query.forEach((value, key) => {
                switch (key) {
                    case WorkspaceProvider.QUERY_PARAM_FOLDER:
                        workspace = { folderUri: workbench_api_1.URI.parse(value) };
                        foundWorkspace = true;
                        break;
                    case WorkspaceProvider.QUERY_PARAM_WORKSPACE:
                        workspace = { workspaceUri: workbench_api_1.URI.parse(value) };
                        foundWorkspace = true;
                        break;
                    case WorkspaceProvider.QUERY_PARAM_EMPTY_WINDOW:
                        workspace = undefined;
                        foundWorkspace = true;
                        break;
                    case WorkspaceProvider.QUERY_PARAM_PAYLOAD:
                        try {
                            payload = JSON.parse(value);
                        }
                        catch (error) {
                            console.error(error);
                        }
                        break;
                }
            });
            if (!foundWorkspace) {
                if (config.folderUri) {
                    workspace = { folderUri: workbench_api_1.URI.revive(config.folderUri) };
                }
                else if (config.workspaceUri) {
                    workspace = { workspaceUri: workbench_api_1.URI.revive(config.workspaceUri) };
                }
            }
            return new WorkspaceProvider(workspace, payload);
        }
        trusted = true;
        constructor(workspace, payload) {
            this.workspace = workspace;
            this.payload = payload;
        }
        async open(workspace, options) {
            if (options?.reuse && !options.payload && this.isSame(this.workspace, workspace)) {
                return true;
            }
            const targetHref = this.createTargetUrl(workspace, options);
            if (targetHref) {
                if (options?.reuse) {
                    window.location.href = targetHref;
                    return true;
                }
                else {
                    return !!window.open(targetHref);
                }
            }
            return false;
        }
        createTargetUrl(workspace, options) {
            let targetHref = undefined;
            if (!workspace) {
                targetHref = `${document.location.origin}${document.location.pathname}?${WorkspaceProvider.QUERY_PARAM_EMPTY_WINDOW}=true`;
            }
            else if ('folderUri' in workspace) {
                const queryParamFolder = encodeURIComponent(workspace.folderUri.toString(true));
                targetHref = `${document.location.origin}${document.location.pathname}?${WorkspaceProvider.QUERY_PARAM_FOLDER}=${queryParamFolder}`;
            }
            else if ('workspaceUri' in workspace) {
                const queryParamWorkspace = encodeURIComponent(workspace.workspaceUri.toString(true));
                targetHref = `${document.location.origin}${document.location.pathname}?${WorkspaceProvider.QUERY_PARAM_WORKSPACE}=${queryParamWorkspace}`;
            }
            if (options?.payload) {
                targetHref += `&${WorkspaceProvider.QUERY_PARAM_PAYLOAD}=${encodeURIComponent(JSON.stringify(options.payload))}`;
            }
            return targetHref;
        }
        isSame(workspaceA, workspaceB) {
            if (!workspaceA || !workspaceB) {
                return workspaceA === workspaceB;
            }
            if ('folderUri' in workspaceA && 'folderUri' in workspaceB) {
                return this.isEqualURI(workspaceA.folderUri, workspaceB.folderUri);
            }
            if ('workspaceUri' in workspaceA && 'workspaceUri' in workspaceB) {
                return this.isEqualURI(workspaceA.workspaceUri, workspaceB.workspaceUri);
            }
            return false;
        }
        isEqualURI(a, b) {
            return a.scheme === b.scheme && a.authority === b.authority && a.path === b.path;
        }
    }
    class LocalStorageURLCallbackProvider {
        _callbackRoute;
        static REQUEST_ID = 0;
        static QUERY_KEYS = [
            'scheme',
            'authority',
            'path',
            'query',
            'fragment'
        ];
        _onCallback = new workbench_api_1.Emitter();
        onCallback = this._onCallback.event;
        pendingCallbacks = new Set();
        lastTimeChecked = Date.now();
        checkCallbacksTimeout = undefined;
        onDidChangeLocalStorageDisposable;
        constructor(_callbackRoute) {
            this._callbackRoute = _callbackRoute;
        }
        create(options = {}) {
            const id = ++LocalStorageURLCallbackProvider.REQUEST_ID;
            const queryParams = [`vscode-reqid=${id}`];
            for (const key of LocalStorageURLCallbackProvider.QUERY_KEYS) {
                const value = options[key];
                if (value) {
                    queryParams.push(`vscode-${key}=${encodeURIComponent(value)}`);
                }
            }
            if (!(options.authority === 'vscode.github-authentication' && options.path === '/dummy')) {
                const key = `vscode-web.url-callbacks[${id}]`;
                localStorage.removeItem(key);
                this.pendingCallbacks.add(id);
                this.startListening();
            }
            return workbench_api_1.URI.parse(window.location.href).with({ path: this._callbackRoute, query: queryParams.join('&') });
        }
        startListening() {
            if (this.onDidChangeLocalStorageDisposable) {
                return;
            }
            const fn = () => this.onDidChangeLocalStorage();
            window.addEventListener('storage', fn);
            this.onDidChangeLocalStorageDisposable = { dispose: () => window.removeEventListener('storage', fn) };
        }
        stopListening() {
            this.onDidChangeLocalStorageDisposable?.dispose();
            this.onDidChangeLocalStorageDisposable = undefined;
        }
        async onDidChangeLocalStorage() {
            const ellapsed = Date.now() - this.lastTimeChecked;
            if (ellapsed > 1000) {
                this.checkCallbacks();
            }
            else if (this.checkCallbacksTimeout === undefined) {
                this.checkCallbacksTimeout = setTimeout(() => {
                    this.checkCallbacksTimeout = undefined;
                    this.checkCallbacks();
                }, 1000 - ellapsed);
            }
        }
        checkCallbacks() {
            let pendingCallbacks;
            for (const id of this.pendingCallbacks) {
                const key = `vscode-web.url-callbacks[${id}]`;
                const result = localStorage.getItem(key);
                if (result !== null) {
                    try {
                        this._onCallback.fire(workbench_api_1.URI.revive(JSON.parse(result)));
                    }
                    catch (error) {
                        console.error(error);
                    }
                    pendingCallbacks = pendingCallbacks ?? new Set(this.pendingCallbacks);
                    pendingCallbacks.delete(id);
                    localStorage.removeItem(key);
                }
            }
            if (pendingCallbacks) {
                this.pendingCallbacks = pendingCallbacks;
                if (this.pendingCallbacks.size === 0) {
                    this.stopListening();
                }
            }
            this.lastTimeChecked = Date.now();
        }
        dispose() {
            this._onCallback.dispose();
        }
    }
    (function () {
        const configElement = window.document.getElementById('vscode-workbench-web-configuration');
        const configElementAttribute = configElement ? configElement.getAttribute('data-settings') : undefined;
        if (!configElement || !configElementAttribute) {
            throw new Error('Missing web configuration element');
        }
        const config = JSON.parse(configElementAttribute);
        (0, workbench_api_1.create)(window.document.body, {
            ...config,
            workspaceProvider: WorkspaceProvider.create(config),
            urlCallbackProvider: new LocalStorageURLCallbackProvider(config.callbackRoute)
        });
    })();
});
